Cleaning up DDP SPMD in reducer.cpp

jaceyca opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65380
* __->__ #64113

Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

jaceyca committed:
Cleaning up DDP SPMD in reducer.cpp

Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

Test Plan: TODO

Reviewers: shenli

Subscribers:

Tasks: T98061288

Tags:

[ghstack-poisoned]

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

Test Plan: TODO

Reviewers: shenli

Subscribers:

Tasks: T98061288

Tags:

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -71,7 +71,7 @@ C10_REGISTER_TYPED_CLASS(TimerRegistry, c10::kCPU, CpuTimer);
 
 Reducer::Reducer(
-    std::vector<std::vector<at::Tensor>> replicas,
+    std::vector<at::Tensor> replicas,
```
The var name `replicas` was referring to replicas of model parameters. Since now there is only one replica, I think we can rename this var to `params`, where which tensor is a parameter.

jaceyca replied:
Done

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -137,9 +135,9 @@ Reducer::Reducer(
   // we do not expect sparse gradients for any parameter.
   if (expect_sparse_gradients_.empty()) {
     expect_sparse_gradients_ = std::vector<std::vector<bool>>(
```
Can be done in followup PRs: Is it possible to simplify this to an `std::vector<bool>`?

jaceyca replied:
Done

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -155,15 +153,15 @@ Reducer::Reducer(
   {
-    const auto replica_count = replicas_.size();
+    const auto replica_count = 1;
     grad_accumulators_.resize(replica_count);
```
can be done in followup PR: can `grad_accumulators_` also be simplified to `std::vector<std::shared_ptr<torch::autograd::Node>>` and get rid of `replica_count `?

jaceyca replied:
Done

mrshenli reviewed (commented):
Hey @jaceyca, thanks for improving Reducer implementation. The build errors are real. Could you please fix those? Thanks!

```
Aug 27 22:04:07 /var/lib/jenkins/workspace/torch/csrc/distributed/c10d/logger.cpp:77:35: error: no matching function for call to 'at::Tensor::size() const'
Aug 27 22:04:07        reducer_->replicas_[0].size();
```

rohan-varma commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -137,9 +135,9 @@ Reducer::Reducer(
+        1, std::vector<bool>(replicas_.size(), false));
   }
-  TORCH_INTERNAL_ASSERT(expect_sparse_gradients_.size() == replicas_.size());
+  TORCH_INTERNAL_ASSERT(expect_sparse_gradients_.size() == 1);
```
This should imply that `expect_sparse_gradients` should now be able to just be a single boolean right?

jaceyca replied:
Yep, I wasn't sure if the related nested containers (expect_sparse_gradients, grad_accumulators) should also be changed

rohan-varma commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -292,27 +290,26 @@ bool Reducer::ddp_graph_static() {
-  }
+  // Deliberately don't pin the memory even if local_used_maps_dev_ will
+  // be cuda. See Note [local_used_maps_ -> local_used_maps_dev copying]
+  local_used_maps_[replica_index] =
```
Probably out of scope for this PR, but in the future `local_used_maps_` can also be refactored to not have to be indexed.

rohan-varma reviewed (commented):
I think appropriate changes also likely need to be made to `c10d/init.cpp` which is the pybind layer for calling C++ code from python, as well as `nn/parallel/distributed.py` where the `Reducer` is used from python.

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

Test Plan: TODO

Reviewers: shenli

Subscribers:

Tasks: T98061288

Tags:

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca commented:
@jaceyca has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30615965).

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

Test Plan: TODO

Reviewers: shenli

Subscribers:

Tasks: T98061288

Tags:

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

Test Plan: TODO

Reviewers: shenli

Subscribers:

Tasks: T98061288

Tags:

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -688,19 +688,17 @@ def __setstate__(self, state):
     def _build_params_for_reducer(self):
         # Build tuple of (module, parameter) for all parameters that require grads.
         modules_and_parameters = [
```
We might need to update the subsequent usage on `modules_and_parameters` as well. Below are a few, there might be more.

https://github.com/pytorch/pytorch/blob/5ecb966e0ff383d65531c8f6de23e704b9cafc54/torch/nn/parallel/distributed.py#L709-L714

change to:

```python
modules_and_parameters = [
    (m, p) for m, p in modules_and_parameters 
    if p not in memo and not memo.add(p)
]
```

https://github.com/pytorch/pytorch/blob/5ecb966e0ff383d65531c8f6de23e704b9cafc54/torch/nn/parallel/distributed.py#L717-L720

change to:

```python
parameters = list(parameter for _, parameter in modules_and_parameters)
```

https://github.com/pytorch/pytorch/blob/5ecb966e0ff383d65531c8f6de23e704b9cafc54/torch/nn/parallel/distributed.py#L732-L735

change to:

```python
expect_sparse_gradient = list(produces_sparse_gradient(module) for module, _ in modules_and_parameters)
```

jaceyca replied:
Made all of these changes, thanks @mrshenli!

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

Test Plan: TODO

Reviewers: shenli

Subscribers:

Tasks: T98061288

Tags:

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

Test Plan: TODO

Reviewers: shenli

Subscribers:

Tasks: T98061288

Tags:

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca commented:
@jaceyca has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30615965).

jaceyca commented on test/distributed/test_c10d_gloo.py:
```diff
@@ -2113,8 +2114,8 @@ def test_reducer_no_multi_replicas(self):
         with self.assertRaisesRegex(
-            RuntimeError,
-            "Expected exactly one model replica.",
+            AssertionError,
```
Wasn't sure if I should modify this test in some other way or just delete it

mrshenli replied:
I'd vote for removing this test, as `Reducer` ctor no longer takes list of lists.

mrshenli commented on torch/csrc/distributed/c10d/reducer.hpp:
```diff
@@ -604,9 +604,9 @@ compute_bucket_assignment_by_size(
+TORCH_API void verify_params_across_processes(
     c10::intrusive_ptr<c10d::ProcessGroup> process_group,
-    std::vector<std::vector<at::Tensor>> model_replicas);
+    std::vector<at::Tensor> model_replica);
```
ditto?

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -2075,13 +2055,13 @@ compute_bucket_assignment_by_size(
+void verify_params_across_processes(
     c10::intrusive_ptr<c10d::ProcessGroup> process_group,
-    std::vector<std::vector<at::Tensor>> model_replicas) {
+    std::vector<at::Tensor> model_replica) {
```
nit: `model_replica` -> `params`?

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1552,10 +1552,10 @@ Example::
+      &::c10d::verify_params_across_processes,
       py::arg("process_group"),
-      py::arg("replicas"),
+      py::arg("model_replica"),
```
nit: `model_replica` -> `params`

mrshenli reviewed (approved):
LGTM! Thanks for this comprehensive revamp!

mrshenli reviewed (commented):
one minor comment: could you please remove the internal-only stuff (eg., test plan, task, review, etc.) from the the PR description? `Differential Revision: D30615965` can stay, as that's the portal to the internal diff. Thanks!

zhaojuanmao commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -292,27 +277,26 @@ bool Reducer::ddp_graph_static() {
 void Reducer::initialize_local_used_map() {
-  const auto replica_count = replicas_.size();
-  const auto variable_count = replicas_[0].size();
+  const auto replica_count = 1;
```
nit: remove this?

jaceyca replied:
I originally left it to keep the appropriate context for lines 282-283, but I can remove it if you think it is cleaner

zhaojuanmao commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -292,27 +277,26 @@ bool Reducer::ddp_graph_static() {
-  const auto replica_count = replicas_.size();
-  const auto variable_count = replicas_[0].size();
+  const auto replica_count = 1;
+  size_t params_index = 0;
```
why is params_index set as 0?

jaceyca replied:
Same reason that replica_count is set to 1, so that we can keep track of how the instances of indexing are related to each other. I can remove this

zhaojuanmao commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -292,27 +277,26 @@ bool Reducer::ddp_graph_static() {
-  }
+  // Deliberately don't pin the memory even if local_used_maps_dev_ will
+  // be cuda. See Note [local_used_maps_ -> local_used_maps_dev copying]
+  local_used_maps_[params_index] =
+      at::zeros({static_cast<long>(variable_count)}, options);
+
+  // This tensor needs to be on the same device as the replica params because
+  // backend such as NCCL may not support CPU tensors, and hence it might not
+  // work if we always put it on CPU.
+  options = options.device(params_[0].device());
+  local_used_maps_dev_[params_index] =
+      at::empty({static_cast<long>(variable_count)}, options);
```
we should not use params_index here?

zhaojuanmao reviewed (commented):
torch/csrc/distributed/c10d/logger.cpp also needs clean up. 

some minor comments

zhaojuanmao commented:
that is weird, all unit tests are passed but logger is not cleaned up.  looks like test_ddp_logging_data_cpu was disabled, I will look into this test. Meanwhile, could you please run a internal benchmark to sanity check it before landing? as it touches many core files of DDP.

Thanks for your effort, great work!

zhaojuanmao commented:
looks like all the related distributed tests are skipped, see https://github.com/pytorch/pytorch/pull/64113/checks?check_run_id=3500595168, 

Determination is skipping distributed/test_c10d_common
Determination is skipping distributed/test_c10d_gloo
Determination is skipping distributed/test_c10d_nccl
Determination is skipping distributed/test_c10d_spawn_gloo
Determination is skipping distributed/test_c10d_spawn_nccl
Running distributed/test_data_parallel without determination
Determination is skipping distributed/test_distributed_spawn
Determination is skipping distributed/test_jit_c10d

mrshenli reviewed (commented):
Looks like `test_ddp_logging_data_cpu` is also skipped in other PRs even though they didn't touch DDP: https://github.com/pytorch/pytorch/pull/63881/checks?check_run_id=3550198101

Some of the distributed tests are covered in this one https://github.com/pytorch/pytorch/pull/64113/checks?check_run_id=3500549588

But looks like it skipped all multi-GPU ones. @walterddr @rohan-varma is this expected? Is there any multi-GPU distributed tests on PRs? And what's the current convention to kick off ci-all master tests? Is it still creating a `ci-all/jaceyca` branch and push from there?

rohan-varma commented:
This isn't expected, we should be running these tests and not sure why the test determinator did not pick these up properly (cc @malfet , @seemethere). I can have a quick patch to remove distributed from target determinator for now.

walterddr commented:
> I can have a quick patch to remove distributed from target determinator for now. I think to run all tests, you can add the "ci/all" label to this PR.

yes i think it is definitely TD that removed distributed/test_distributed_spawn from the list based on the log. if I remember correctly the label to be added is `ci/master`

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

mrshenli reviewed (commented):
Thanks @rohan-varma and @walterddr for the quick turnaround!

Hey @jaceyca, could you please rebase when #64721 is landed? Thanks!

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

rohan-varma commented:
It looks like `test_ddp_logging_data_cpu` ran and passed on https://github.com/pytorch/pytorch/pull/64113/checks?check_run_id=3562753121

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -292,27 +277,24 @@ bool Reducer::ddp_graph_static() {
-        at::empty({static_cast<long>(variable_count)}, options);
-  }
+  const auto variable_count = params_.size();
+  local_used_maps_.resize(1);
+  local_used_maps_dev_.resize(1);
```
could we define 'at::Tensor local_used_maps_' and 'at::Tensor local_used_maps_dev_'?

mrshenli replied:
The name needs to be changed to `at::Tensor local_used_map_` and `at::Tensor local_used_map_dev_`? (maps -> map)

jaceyca replied:
If I change the typing of these variables, I will also have to change their usages. Just to be clear, is that what you're asking for?

jaceyca replied:
Addressed in https://github.com/pytorch/pytorch/pull/65380

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca commented:
@jaceyca has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30615965).

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca commented:
@jaceyca has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30615965).

jaceyca commented:
`test/distributed/test_c10d_gloo.py::ProcessGroupGlooTest::test_allgather_coalesced_async` is failing but I don't think it's related to my PR since it fails on master as well.

mrshenli commented:
> test/distributed/test_c10d_gloo.py::ProcessGroupGlooTest::test_allgather_coalesced_async is failing but I don't think it's related to my PR since it fails on master as well.

That's my bad. Let me disable that.

mrshenli commented:
Hey @jaceyca, could you please point me to the master test where `test_allgather_coalesced_async` fails?

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca commented:
@jaceyca has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30615965).

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca commented:
@jaceyca has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30615965).

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca commented:
@jaceyca has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30615965).

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca commented:
@jaceyca has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30615965).

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1566,20 +1566,28 @@ Example::
   module.def(
       "_verify_model_across_ranks",
```
curious, is it possible to consolidate `_verify_model_across_ranks` with `_verify_params_across_processes`?

rohan-varma replied:
Yeah, seems that we can just delete this as we've renamed it?

jaceyca replied:
Done

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca committed:
Update on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca commented:
@jaceyca has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30615965).

jaceyca committed:
Rebase on "Cleaning up DDP SPMD in reducer.cpp"


Summary: Since there is only one model replica per process, `replicas`
can be simplified from `std::vector<std::vector<at::Tensor>>` to
`std::vector<at::Tensor>` in the Reducer class.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Differential Revision: [D30615965](https://our.internmc.facebook.com/intern/diff/D30615965)

[ghstack-poisoned]

jaceyca commented:
@jaceyca has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30615965).
